File: src/collectors/homebrew.ts
import type { Collector, CollectorResult } from "./types";
import { makeSection, toItems } from "./types";
import { type CommandEnv, defaultEnv } from "./env";

/** Parse `brew list --formula` / `--cask` (one name per line). */
export function parseBrewList(text: string): string[] {
  return text
    .trim()
    .split("\n")
    .map((l) => l.trim())
    .filter(Boolean)
    .sort();
}

// Progress/noise lines `brew bundle dump` may emit (denylist). Everything else is kept,
// so first-class directives (go/npm/cargo/uv/whalebrew/vscode/mas/…) survive — an
// allowlist silently dropped go/npm and produced an incomplete "restorable" Brewfile.
const BREWFILE_NOISE = /^\s*(✔|✓|✗|⚠|ℹ|==>|Warning:|Error:)|JSON API/;

/** Clean `brew bundle dump` stdout into a restorable Brewfile (drop noise, keep every directive). */
export function parseBrewfile(text: string): string {
  return text
    .split("\n")
    .filter((l) => !BREWFILE_NOISE.test(l))
    .join("\n")
    .trim();
}

export function makeHomebrewCollector(env: CommandEnv = defaultEnv): Collector {
  return async () => {
    const result: CollectorResult = {};

    try {
      const formulae = parseBrewList(await env.run(["brew", "list", "--formula"]));
      if (formulae.length)
        result["apps.brew.formulae"] = makeSection("apps.brew.formulae", { items: toItems(formulae) });
    } catch {}

    try {
      const casks = parseBrewList(await env.run(["brew", "list", "--cask"]));
      if (casks.length) result["apps.brew.casks"] = makeSection("apps.brew.casks", { items: toItems(casks) });
    } catch {}

    // A restorable Brewfile (taps + brews + casks + mas) — superset used by `brew bundle`.
    try {
      const brewfile = parseBrewfile(await env.run(["brew", "bundle", "dump", "--file=-"]));
      if (brewfile) result["apps.brew.bundle"] = makeSection("apps.brew.bundle", { content: brewfile });
    } catch {}

    return result;
  };
}

export const collectHomebrew = makeHomebrewCollector();

File: internal/collect/homebrew.go
package collect

import (
	"regexp"
	"sort"
	"strings"

	"github.com/doguyilmaz/dothaven/internal/snapshot"
)

// brewfileNoise matches progress/noise lines that `brew bundle dump` may emit on
// a cold cache. Everything else is kept, so first-class directives
// (go/npm/cargo/uv/whalebrew/vscode/mas/…) survive — an allowlist silently
// dropped go/npm and produced an incomplete "restorable" Brewfile.
var brewfileNoise = regexp.MustCompile(`^\s*(✔|✓|✗|⚠|ℹ|==>|Warning:|Error:)|JSON API`)

// ParseBrewList parses `brew list --formula` / `--cask` (one name per line):
// trims, drops blanks, sorts.
func ParseBrewList(text string) []string {
	out := []string{}
	for _, line := range strings.Split(strings.TrimSpace(text), "\n") {
		if l := strings.TrimSpace(line); l != "" {
			out = append(out, l)
		}
	}
	sort.Strings(out)
	return out
}

// ParseBrewfile cleans `brew bundle dump` stdout into a restorable Brewfile:
// drops noise lines, keeps every directive, trims surrounding blank lines.
func ParseBrewfile(text string) string {
	var kept []string
	for _, line := range strings.Split(text, "\n") {
		if !brewfileNoise.MatchString(line) {
			kept = append(kept, line)
		}
	}
	return strings.TrimSpace(strings.Join(kept, "\n"))
}

// HomebrewCollector runs `brew list`/`brew bundle dump` and emits the installed
// formulae, casks, and a restorable Brewfile.
func HomebrewCollector(c Ctx) snapshot.Snapshot {
	out := snapshot.Snapshot{}

	formulaeOut, _ := c.Env.Run(c.Context, "brew", "list", "--formula")
	if formulae := ParseBrewList(formulaeOut); len(formulae) > 0 {
		out["apps.brew.formulae"] = snapshot.Section{Items: toItems(formulae)}
	}

	casksOut, _ := c.Env.Run(c.Context, "brew", "list", "--cask")
	if casks := ParseBrewList(casksOut); len(casks) > 0 {
		out["apps.brew.casks"] = snapshot.Section{Items: toItems(casks)}
	}

	bundleOut, _ := c.Env.Run(c.Context, "brew", "bundle", "dump", "--file=-")
	if bundle := ParseBrewfile(bundleOut); bundle != "" {
		out["apps.brew.bundle"] = snapshot.Section{Content: ptr(bundle)}
	}

	return out
}
